feat(jwt): remote jwks, remote signing, and exportable signJwt function

dvanmali opened the pull request:
## Features

**feat**: `remoteUrl` option disables jwks endpoint and uses this endpoint in oAuth metadata

**feat**: remote `sign` payloads using a server-side privateKey environment variable or remote signing service like Google KMS, AWS KMS, Azure Key Vault, etc.

**feat**: `signJwt` (previously `getJwtToken`) now accepts payload as a parameter. Those who utilized exported `getJwtToken` should utilize `signJwt` and its payload parameter instead of through options. `getJwtToken` marked deprecated. `signJwt` now has safeguards that is enabled by default and can be disabled per plugin implementation.

**feat**: Added `disableSettingJwtHeader` to disable setting jwt in header. Useful for oAuth compliance.

**feat**: exports `getJwtPlugin` for other plugins

**chore**: combine shared createJwks functionality

__Duplicate #3464__
__Partial #3458__

## Breaking Changes

**MINOR**
- style: `JwtOptions` renamed to `JwtPluginOptions` to match other plugin syntax.
- fix: `getJwtToken` deprecated, prefer `signJwt` instead

dvanmali committed:
feat: remote jwks, remote signing, and exportable signJwt function

feat: remoteUrl disablesjwks endpoint and uses this endpoint in discovery mode

feat: remote sign payloads to such as Google KMS, AWS KMS, and Azure Key Vault

fix: signJwt now accepts payload as a parameter

LightTab2 reviewed (changes requested):
I have some security concerns and think some safeguards could be used to make sure user  wants to override critical settings with their payload. I can imagine a case when someone creates a sloppy API endpoint that signs anything and the payload could be maliciously crafted

dvanmali commented on packages/better-auth/src/plugins/jwt/sign.ts:
```diff
@@ -53,23 +60,101 @@ export async function getJwtToken(
-		.setSubject(
-			(await options?.jwt?.getSubject?.(ctx.context.session!)) ??
-				ctx.context.session!.user.id,
+		.setIssuedAt(payload.iat)
```
JWT structure is `${protectedHeader}.${jwtPayload}.${signature}`. Typically the payload should never have `alg` defined as this is entirely determined by the config.

If you put `alg` then technically the output would be in a payload claim not a header claim. In which verification from packages like `jose` aren't designed to verify. Plus it'll be redundant if defined twice.

dvanmali replied:
For these, if these are defined inside payload, it'll work in the same way as setting them individually.

dvanmali committed:
style: change disallow to allow for readability

dvanmali committed:
fix: make clear the exportable functions of plugin

dvanmali committed:
feat: add iat to consolidate start time

dvanmali committed:
fix: iat and exp should be in seconds but not reflected

dvanmali commented:
@cubic-dev-ai review again

dvanmali commented:
@cubic-dev-ai review again

LightTab2 commented on packages/better-auth/src/plugins/jwt/utils.ts:
```diff
@@ -1,5 +1,83 @@
+};
+
+export function toExpJWT(input: number | Date | string, iat?: number): number {
+	const now = iat ?? Math.floor(Date.now() / 1000); // current Unix timestamp in seconds
```
Adding `iat` here allows to create correct JWT tokens that are to be send in future, like caching them... It's weird but some might want to use this scenario, however if `iat > Date.now()/1000` and this token gets leaked, somebody basically got their hands on a token that has longer Expiration Time than allowed... So I propose another safeguard:
`allowFutureIatWithoutNbf`. 
Logic of the safeguard would be:
> Set `nbf` to `iat`, if `iat` is set to future AND `nbf `is `undefined` or **lesser** than `iat` unless `allowFutureIatWithoutNbf` is `true`

dvanmali replied:
`signJwt` must at least follow the strict standards of [JWT RFC 7519](https://datatracker.ietf.org/doc/html/rfc7519#section-4) which specifies the `iat`, `exp`, and `nbf` claims. Thus, `signJwt` should only sign a **valid** JWT payload (not just data) makes the most valid descriptive sense to me.

LightTab2 commented on packages/better-auth/src/plugins/jwt/utils.ts:
```diff
@@ -1,5 +1,83 @@
+	const now = iat ?? Math.floor(Date.now() / 1000); // current Unix timestamp in seconds
+
+	if (typeof input === "number") {
+		return now + input;
```
This is wrong, when `exp` is a `number`, it is a **UNIX timestamp**.

LightTab2 commented on packages/better-auth/src/plugins/jwt/sign.ts:
```diff
@@ -55,23 +108,120 @@ export async function getJwtToken(
+		payload.sub ??
+		(await options?.jwt?.getSubject?.(ctx.context.session!)) ??
+		ctx.context.session?.user.id;
+	if (sub) jwt.setSubject(sub);
```
```js
	if (payload.nbf) jwt.setNotBefore(payload.nbf);
	if (payload.jti) jwt.setJti(payload.jti);
```
Any reason to not set these?

LightTab2 commented on packages/better-auth/src/plugins/jwt/sign.ts:
```diff
@@ -55,23 +108,120 @@ export async function getJwtToken(
+ * @param ctx - endpoint context
+ */
+// Plugin exportable
+export async function createJwk(ctx: GenericEndpointContext) {
```
Note lack of `options` as breaking change.

dvanmali replied:
This code is a new function entirely. Not breaking. Equivalent to `signJwt` in which no options is provided and uses the option. Intended use is for a future admin plugin to be able to create/rotate jwks, thus is exported in index.ts

LightTab2 commented on packages/better-auth/src/plugins/jwt/utils.ts:
```diff
@@ -1,5 +1,83 @@
+	return Math.floor(isSubtraction ? now - totalSeconds : now + totalSeconds);
+}
 
 async function deriveKey(secretKey: string): Promise<CryptoKey> {
```
Are `deriveKey`, `encryptPrivateKey`, `decryptPrivateKey` functions used anywhere?

dvanmali replied:
Looking at the history, it appears the dependency both moved away into `better-auth/utils` package and changed down the road to use the `jose` package and its `generateKeyPair`.

LightTab2 commented on packages/better-auth/src/plugins/jwt/sign.ts:
```diff
@@ -1,42 +1,95 @@
+	// Exp safety check
+	let exp = payload.exp;
+	const defaultExp = options?.jwt?.expirationTime ?? "1h";
+	if (disallowLargerExpTime && exp && exp > toExpJWT(defaultExp, iat)) {
```
Correct to
```js
if (disallowLargerExpTime && (!exp || exp > toExpJWT(defaultExp, iat)))
```

Missing `exp` creates an infinite token, can't have that. In fact there should be a comment as precaution that `disallowLargerExpTime` will allow to create infinite tokens.

dvanmali replied:
Shouldn't create infinite token. It'll go to the default.

LightTab2 commented on packages/better-auth/src/plugins/jwt/sign.ts:
```diff
@@ -1,42 +1,95 @@
+		payload = {
+			...payload,
+			iat,
+			exp: exp ?? toExpJWT(defaultExp, iat),
```
Don't know if other params are needed. Honestly, I don't know if anything related to remote signing could be improved upon, because I would have to try using it in the first place. Someone could take a look at this too.

dvanmali replied:
I don't believe the jwt function needs to verify anything beyond `iat`, `exp`, `iss`, and `aud`. The payload in oidc plugin will be able to add custom claims to `id_token` and JWT `access_tokens` in which verification of those claims should be completed in oidc plugin.

LightTab2 reviewed (changes requested):
`exp` check needs to be fixed. Other comments are not so important and opinion-biased.

dvanmali committed:
fix: suggestions from LightTab2

fix: guarantee iat and exp in jwt claims

feat: nbf checking

chore: remove unused utils

dvanmali commented on packages/better-auth/src/plugins/jwt/types.ts:
```diff
@@ -0,0 +1,159 @@
+	 * subtracted from the current unix timestamp. A "from now" suffix can also be used for
+	 * readability when adding to the current unix timestamp.
+	 *
+	 * @default 1hr
```
@LightTab2 if this sounds good to you, you or I can mark this as resolved.

dvanmali replied:
I agree access tokens are the most secure with lower expiration times at `15 min`, which should be used in highly secure systems. But `1 hr` works best for most web applications and this setting serves as a default absolute max for other plugins. Other plugins can always lower their absolute expiration (these JWT plugin setting) or manually lower the exp field in their plugin (usually scope dependent). There's just a lot more breaking changes and higher likeliness to setup improper settings if `15 min` is kept.

dvanmali committed:
fix: remove allowFutureIatTime flag, iat shall always be current or in the past, and iat is optional

dvanmali commented on packages/better-auth/src/plugins/jwt/sign.ts:
```diff
@@ -1,42 +1,100 @@
+
+	// Nbf strict checks
+	const nbf = payload.nbf;
+	if (nbf &&  !((iat && nbf < iat) || nbf > exp)) {
```
Implementation Note: Using `iat && nbf < iat` instead of `nbf < (iat ?? nowSeconds)` because nbf can only be checked against iat if sent, nowSeconds is local and may fail if used in this comparison.

dvanmali commented on packages/better-auth/src/plugins/jwt/sign.ts:
```diff
@@ -1,42 +1,100 @@
+	const nowSeconds = Math.floor(Date.now() / 1000);
+	const iat = payload.iat;
+	if (iat && iat > nowSeconds) {
+		throw new BetterAuthError("unable to set a future iat time");
```
By definition `iat` is the current issued time. It can never be set to a future time nor should it be bypassed. Utilize `nbf` if the issuance is set for the future.

dvanmali commented:
@cubic-dev-ai review again

LightTab2 reviewed (commented):
I didn't review OIDC changes; custom sign function tries to achieve golden ratio between customization and safety, personally I'd like to have 2 separate functions, one which simply signs using plugin's JWT settings and one which allows for every shenanigans possible (with togglable safeguards), but this is a good solution all in all and here I talked only about my preferences.

I won't mark this review as approved, because I want somebody more knowledgable to take notion of the default time change from "15m" to "1h" and decide whether its a good move first and if we need to make a workaround around OIDC instead.

dvanmali commented:
Ready for review @Bekacru

dvanmali committed:
fix: remote plugin checking

dvanmali committed:
fix: cubic-dev-ai suggestions

dvanmali committed:
chore: move back into return statement to enable plugin users to prevent client import warning

dvanmali committed:
fix: remove requirement of plugin order

dvanmali commented:
@Bekacru?

Bekacru commented on packages/better-auth/src/plugins/jwt/index.ts:
```diff
@@ -321,23 +180,59 @@ export const jwt = (options?: JwtOptions) => {
 				async (ctx) => {
-					const jwt = await getJwtToken(ctx, options);
+					// Disables endpoint if using oAuth provider
+					if (options?.usesOAuthProvider) {
```
I'm not sure if see the point of using this flag. The goal seems to be only to disable the token endpoint or am I missing something? If that's the case, they can already do that in other ways, including with the `disabledPaths` option.

dvanmali replied:
NOTE: this requires all `oidcProvider` plugins to set `allowAudienceMismatch: true` in order to sign id tokens

dvanmali replied:
Was able to remove `usesOAuthProvider` by adding a `safeguards` parameter instead of going through options.

dvanmali commented on packages/better-auth/src/plugins/jwt/index.ts:
```diff
@@ -321,23 +180,59 @@ export const jwt = (options?: JwtOptions) => {
 					},
 					handler: createAuthMiddleware(async (ctx) => {
+						// Disables middleware if using oAuth provider
+						if (options?.usesOAuthProvider) {
```
Valid, will implement

dvanmali committed:
fix: minor Bekacru suggestions

dvanmali committed:
fix(jwt): remove usesOAuthProvider flag

fix(oidc): requires setting allowAudienceMismatch to true

dvanmali commented:
@cubic-dev-ai review

dvanmali commented on packages/better-auth/src/plugins/jwt/sign.ts:
```diff
@@ -1,42 +1,103 @@
+	const aud = payload.aud;
+	const defaultAud = options?.jwt?.audience ?? ctx.context.options.baseURL!;
+	const allowAudienceMismatch = options?.jwt?.allowAudienceMismatch;
+	if (!allowAudienceMismatch && aud) {
```
Flag was removed and linked plugin should set `allowAudienceMismatch` instead

dvanmali committed:
fix: suggestion from cubic-dev-ai

dvanmali committed:
chore: move sign safety checks into a parameter instead of option

dvanmali committed:
fix: revert expirationTime to 15m default

dvanmali commented on docs/content/docs/plugins/jwt.mdx:
```diff
@@ -304,3 +304,104 @@ jwt({
+
+### OAuth Provider Mode
+
+If you are making your system oAuth compliant (such as when utilizing the OIDC or MCP plugins), you **MUST** disable the `/token` endpoint (oAuth equivalent `/oauth2/token`) and disable setting the jwt header (oAuth equivalent `oauth2/userinfo`).
```
It should be a must because oAuth requires scope checking. This jwt plugin does not require scope checking.

dvanmali commented on packages/better-auth/src/plugins/jwt/sign.ts:
```diff
@@ -55,23 +116,123 @@ export async function getJwtToken(
+/**
+ * Backwards compatable version of signJwt
+ *
+ * @deprecated - prefer signJwt to prevent option injection
```
Actually does not prevent option injection because updating the `ctx.context` directly would do the same.

gustavovalverde commented:
@dvanmali this was partially shipped in `next`, I'm closing this PR, but if you'd like to continue with the the work please open a new updated PR
